aidb/inference/bound_inference_service.py

```diff
@@ -170,22 +169,27 @@ async def _insert_output_results_in_tables(self, output_results_data: List[pd.Da
         await conn.run_sync(lambda conn: tmp_df.to_sql(table, conn, if_exists='append', index=False))
 
 
-  async def _check_inputs_in_cache_table(self, inputs: pd.DataFrame, conn):
+  async def _get_inputs_not_in_cache_table(self, inputs: pd.DataFrame, conn):
     """
-    checks the presence of inputs in the cache table
+    Checks the presence of inputs in the cache table and returns the inputs that are not in the cache table.
     """
     cache_entries = await conn.run_sync(lambda conn: pd.read_sql_query(text(str(self._cache_query_stub.compile())), conn))
     cache_entries = cache_entries.set_index([col.name for col in self._cache_columns])
-    normalized_cache_cols = [self.convert_cache_column_name_to_normalized_column_name(col.name) for col in
-                             self._cache_columns]
-    is_in_cache = []
+    normalized_cache_cols = [self.convert_cache_column_name_to_normalized_column_name(col.name) for col in self._cache_columns]
+
     if len(normalized_cache_cols) == 1:
-      for ind, row in inputs.iterrows():
-        is_in_cache.append(row[normalized_cache_cols[0]] in cache_entries.index)
+        # For a single column, use `isin` and negate the condition
+        col = normalized_cache_cols[0]
+        out_cache_df = inputs[~inputs[col].isin(cache_entries.index)]
+        out_cache_df_primary = out_cache_df[[col]]
     else:
-      for ind, row in inputs.iterrows():
-        is_in_cache.append(tuple([row[col] for col in normalized_cache_cols]) in cache_entries.index)
-    return is_in_cache
+        # For multiple columns, create tuples and use set operations
+        inputs_tuples = inputs[normalized_cache_cols].apply(tuple, axis=1)
+        cache_tuples = set(cache_entries.index)
+        out_cache_df = inputs[~inputs_tuples.isin(cache_tuples)]
+        out_cache_df_primary = out_cache_df[normalized_cache_cols]
```

Comment:
I've modified the cache table to only contain the primary key columns. Is there any difference between out_cache_df and out_cache_df_primary?